include/vast/Dialect/HighLevel/HighLevelOps.td

```diff
@@ -633,28 +667,50 @@ class ArithBinOp< string mnemonic, list< Trait > traits = [] >
 
         The custom assembly form of the operation is as follows:
 
-        %result = <op> %lhs, %rhs  : type
+        %result = <op> %lhs, %rhs  : functional-type(operands, results)
     }];
 
-    let assemblyFormat = [{ $lhs `,` $rhs attr-dict `:` type($result) }];
+    let skipDefaultBuilders = 1;
+    let builders = [ OpBuilder<(ins "Value":$lhs, "Value":$rhs), [{
+        $_state.addOperands(lhs);
+        $_state.addOperands(rhs);
+        auto type = lhs.getType();
+        if(rhs.getType().isa< hl::PointerType >())
+        {
+            // Both operands are pointers -> it's a subtraction
+            if(type.isa< hl::PointerType >())
+                type = odsBuilder.getType< hl::LongType >();
+            else
+                type = rhs.getType();
+        }
+        $_state.addTypes(type);
+      }]>
+    ];
+
+    let assemblyFormat = [{ $lhs `,` $rhs attr-dict `:` functional-type(operands, results) }];
 }
 
-def AddIOp : ArithBinOp< "add", [Commutative]>;
-def AddFOp : ArithBinOp<"fadd", []>;
-def SubIOp : ArithBinOp< "sub", []>;
-def SubFOp : ArithBinOp<"fsub", []>;
-def MulIOp : ArithBinOp< "mul", [Commutative]>;
-def MulFOp : ArithBinOp<"fmul", []>;
-def DivSOp : ArithBinOp<"sdiv", []>;
-def DivUOp : ArithBinOp<"udiv", []>;
-def DivFOp : ArithBinOp<"fdiv", []>;
-def RemSOp : ArithBinOp<"srem", []>;
-def RemUOp : ArithBinOp<"urem", []>;
-def RemFOp : ArithBinOp<"frem", []>;
+class StandardArithBinOp< string mnemonic, list< Trait > traits = [] >
+    : ArithBinOp< mnemonic, !listconcat(traits, [SameOperandsAndResultType]) >;
+
+
+def AddIOp : ArithBinOp< "add", [Commutative, IsAdditive< "lhs", "rhs", "result" >]>;
```

Comment:
can not you have addition of mixed types?

Reply:
similarly subtraction?

Reply:
I think that in that case the right hand side should be casted to match the types. At least that's what happens everywhere else. Why should it be different here?